### DISCUSSION TITLE
Questions about a personal project (node graph GUI, CLI-API mapping, opinion)

### DISCUSSION THREAD

**Discussion by sseerrppeenntt** [ImageMagick/ImageMagick#7581] [Help]
### ImageMagick version

-

### Operating system, version and so on

-

### Description

Hey friends ❤

I am a novice dev and in my spare time I am playing with @dlemstra's magick-wasm. From my understanding the magick-wasm is a mapping of the C API. 

The idea is a project where one could specify an image pipeline using a node graph (as one can in Blender with nodes), with an input node, an output node and nodes in the middle modifying the image. This would help maintain the technical acuity of using the CLI/API, as opposed to "over-abstracted" GUIs like GIMP or Photoshop. The main use case would be to output images as do any other IM GUIs, but also it would help people to build CLI commands graphically (that would be a secondary output of the program.)

Question 1: Is there some project utilizing ImageMagick with a node graph-like GUI?
_(So I am not reinventing any wheels)_

Illustration 1 (Geometry Nodes in Blender):
![463f19e2593f93ee229b2163a3d37c048880204a](https://github.com/user-attachments/assets/28d46189-d583-46bc-896d-c173c3aa7c06)

_I am sure this is asked a lot, but I couldn't find such post:_
Question 2: Could you explain to me (or point to some resources on) how is the underlying API called when using CLI? Are there perhaps some translators (CLI command <-> API call?)

Question 3: Do you think such a tool would be beneficial?

**Comment by snibgo** [comment]
> Question 1: Is there some project utilizing ImageMagick with a node graph-like GUI?

Not as far as I know.

> Question 2: Could you explain to me (or point to some resources on) how is the underlying API called when using CLI? Are there perhaps some translators (CLI command <-> API call?)

I suggest the source code. In particular, file MagickWand/operation.c function CLISimpleOperatorImage() deals with most operations in a large `case` statement. For example:
```
      if (LocaleCompare("bilateral-blur",option+1) == 0)
        {
          flags=ParseGeometry(arg1,&geometry_info);
          if ((flags & (RhoValue|SigmaValue)) == 0)
            CLIWandExceptArgBreak(OptionError,"InvalidArgument",option,arg1);
          if ((flags & SigmaValue) == 0)
            geometry_info.sigma=geometry_info.rho;
          if ((flags & XiValue) == 0)
            geometry_info.xi=1.0*sqrt(geometry_info.rho*geometry_info.rho+
              geometry_info.sigma*geometry_info.sigma);
          if ((flags & PsiValue) == 0)
            geometry_info.psi=0.25*sqrt(geometry_info.rho*geometry_info.rho+
              geometry_info.sigma*geometry_info.sigma);
          new_image=BilateralBlurImage(_image,(size_t) geometry_info.rho,
            (size_t) geometry_info.sigma,geometry_info.xi,geometry_info.psi,
           _exception);
          break;
        }
```
The CLI option `-bilateral-blur AxB+C+D` (see https://imagemagick.org/script/command-line-options.php#bilateral-blur) calls `BilateralBlurImage(_image,A,B,C,D,exception)` (which is in `MagickWand/effect.c`).

EDIT: This is a common naming pattern: CLI `-blur` calls `BlurImage()`, `-polaroid` calls `PolaroidImage()`, and so on.

Dealing with pseudo-options like `(` and `)` needs a bit of care.

> Question 3: Do you think such a tool would be beneficial?

Yes. I would especially like it if the tool could (a) generate a node graph from a given `magick` command, and (b) the reverse, generate a `magick` command from a node graph.

**Comment by dlemstra** [comment]
The `magick-wasm` library uses the C api of ImageMagick directly. There is no `magick` cmd api available and I will also not be going to add that. In theory you could however translate those api calls back to commands.
